### ISSUE TITLE
Space / time leak

### ISSUE THREAD

**Issue by nickkuk** [HeinrichApfelmus/reactive-banana#152]
The following code leaks for reactive-banana-1.1.0.1 and GHC-8.0.2:
```
module Main where

import Control.Monad
import Reactive.Banana.Frameworks
import Reactive.Banana

main :: IO ()
main = do
  (addHandler, fire) <- newAddHandler
  let nw = do
        e1 <- fromAddHandler addHandler
        e2 <- switchE (never <$ e1)
        reactimate $ return () <$ e2
  compile nw >>= actuate
  forever $ fire ()
```

**Comment by ocharles** [comment]
Could you add an explicit GC after `fire ()` to confirm?

**Comment by nickkuk** [comment]
Do you mean
```
module Main where

import Control.Monad
import System.Mem
import Reactive.Banana.Frameworks
import Reactive.Banana

main :: IO ()
main = do
  (addHandler, fire) <- newAddHandler
  let nw = do
        e1 <- fromAddHandler addHandler
        e2 <- switchE (never <$ e1)
        reactimate $ return () <$ e2
  compile nw >>= actuate
  forever $ do
    fire ()
    performMajorGC
```
? It seems that memory grows too, but much slower.

**Comment by nickkuk** [comment]
@ocharles Or it stop to grow at some value... Does it exists robust way to confirm that there is no leak for long living program?

**Comment by ocharles** [comment]
That's what I meant, I just wanted to verify it was a real leak, and not just GC not kicking in.

**Comment by vagarenko** [comment]
@nickkuk 
There is a way
http://neilmitchell.blogspot.com/2015/09/detecting-space-leaks.html

**Comment by nickkuk** [comment]
@vagarenko It seems that aforementioned code doesn't use any stack. I think that leaks in reactive-banana library (if they're exist) should be more like ones from the last point in "Caveats" section from [http://neilmitchell.blogspot.com/2015/09/detecting-space-leaks.html](http://neilmitchell.blogspot.com/2015/09/detecting-space-leaks.html), i.e. heap leaks.

**Comment by nickkuk** [comment]
@ocharles @HeinrichApfelmus I can confirm that memory consumption grows with performMajorGC too. It starts with 3 MB and eats 200 MB during the night.

**Comment by HeinrichApfelmus** [comment]
Om nom nom. That's unusual and shouldn't happen, I think. 🤔 Unfortunately, it will take a while until I can look into this, I'm currently a bit swamped.

**Comment by DevJac** [comment]
I can confirm this memory leak (has it been confirmed as such?) happens on both GHC 7.10 and GHC 8.0. I am running the test code which contains `performMajorGC`.

**Comment by mitchellwrosen** [comment]
Simple profiling indicates this line is the cause of increasing heap allocation over time: https://github.com/HeinrichApfelmus/reactive-banana/blob/7b120be3667488cfb46f48cef32bf91c5fbbb366/reactive-banana/src/Reactive/Banana/Prim/Dependencies.hs#L50

**Comment by mitchellwrosen** [comment]
I think the problem is the line in that I linked last August. We have a child pulse (the output of `switchE`) repeatedly changing parents. Each corresponding `connectChild` call allocates a new weak reference for the parent (in this case, `never`) to point at the child (`switchE` output). These weak references are never collected because the child pulse is never freed.

Here is a similar program that leaks memory:

```haskell
main = do
  ref <- newIORef 0
  forever $ do
    mkWeakIORef ref (pure ())
    performGC
```

**Comment by mitchellwrosen** [comment]
I looked into this a bit more. No huge update, but that line in `connectChild` above is not the only source of weak references accumulating forever. (I tested this by creating storing a weak ref to oneself in every node, then reusing this when adding a child's weak ref to its parent).

The weak reference that is used to keep a child's parent alive is also never being freed, but I'm not sure why, since this one is `finalize`d when changing parents.
